src/libstore/build/derivation-goal.cc

```diff
@@ -770,35 +771,42 @@ void DerivationGoal::tryLocalBuild() {
 }
 
 
-static void chmod_(const Path & path, mode_t mode)
-{
-    if (chmod(path.c_str(), mode) == -1)
-        throw SysError("setting permissions on '%s'", path);
-}
-
-
 /* Move/rename path 'src' to 'dst'. Temporarily make 'src' writable if
    it's a directory and we're not root (to be able to update the
    directory's parent link ".."). */
 static void movePath(const Path & src, const Path & dst)
 {
-    auto st = lstat(src);
+    namespace fs = std::filesystem;
+    auto st = fs::symlink_status(src);
 
     bool changePerm = (
 #ifndef _WIN32
         geteuid()
 #else
         !isRootUser()
 #endif
-        && S_ISDIR(st.st_mode) && !(st.st_mode & S_IWUSR));
+        && fs::is_directory(st)
+        && (st.permissions() & fs::perms::owner_write) == fs::perms::none);
 
-    if (changePerm)
-        chmod_(src, st.st_mode | S_IWUSR);
+    auto permOpts = fs::perm_options::replace | fs::perm_options::nofollow;
 
-    std::filesystem::rename(src, dst);
+    if (changePerm) {
+        try {
+            fs::permissions(src, st.permissions() | fs::perms::owner_write, permOpts);
+        } catch (const fs::filesystem_error & e) {
+            throw SysError("setting permissions on '%s'", src);
```

Comment:
Does `SysError` actually work correctly after a `std::filesystem` exception? I don't know if it's guaranteed to leave `errno` in a known state.

Reply:
The [documentation](https://en.cppreference.com/w/cpp/filesystem/permissions#Exceptions) says that `fs::filesystem_error` is only thrown on OS errors so I think it's safe to assume that the `errno` is appropriately set when we're in this `catch` block?

Reply:
That's probably a safe assumption on Unix-like systems, but I'm not sure if it's the case on Windows. Maybe @Ericson2314 knows?

Reply:
We can use

-  https://en.cppreference.com/w/cpp/error/system_error/code to get the code
- https://en.cppreference.com/w/cpp/error/error_code/value to get the Unix err code
- https://hydra.nixos.org/build/264900059/download/1/html/classnix_1_1SysError.html#ae7e9f454d2fc92588de15bcef87ec8c4 to construct a `SysError` with an explicitly given `errNo`

Reply:
I suppose on windows we than get these long error codes instead?

Reply:
@Ericson2314 I have changed the code to use the `SysError` with explicit `errNo` (see https://github.com/NixOS/nix/pull/10937/commits/e0090fea40d4e7cdcbfbb879a7507be4f5f5f625)